src/NovelRT/Experimental/Graphics/GraphicsMemoryBlockCollection.cpp

```diff
@@ -0,0 +1,481 @@
+                {
+                    size_t size = GetSize();
+                    size_t minimumSize = GetMinimumSize();
+
+                    // We have two empty blocks, we want to prefer removing the larger of the two.
+
+                    if (block->GetSize() > _emptyBlock->GetSize())
+                    {
+                        if ((size - block->GetSize()) >= minimumSize)
+                        {
+                            RemoveBlockAt(blockIndex);
+                        }
+                        else if ((size - _emptyBlock->GetSize()) >= minimumSize)
+                        {
+                            RemoveBlock(_emptyBlock);
+                        }
+                    }
+                    else if ((size - _emptyBlock->GetSize()) >= minimumSize)
+                    {
+                        RemoveBlockAt(blockIndex);
+                    }
+                    else if ((size - block->GetSize()) >= minimumSize)
+                    {
+                        RemoveBlock(_emptyBlock);
+                    }
+
+                    // Removing either would put us below the minimum size if none of the above is met, so we can't
+                    // remove.
+                }
+                else if (block->GetSize() > _emptyBlock->GetSize())
+                {
+                    _emptyBlock = block;
+                }
+            }
+            else
+            {
+                // We have no existing empty blocks, so we want to set the index to this block unless
+                // we are currently exceeding our memory budget, in which case we want to free the block
+                // instead. However, we still need to maintain the minimum block count and minimum size
+                // placed on the collection, so we will only respect the budget if those can be maintained.
+
+                GraphicsMemoryBudget budget =
+                    _allocator->GetBudget(std::static_pointer_cast<GraphicsMemoryBlockCollection>(shared_from_this()));
+
+                if ((budget.GetEstimatedUsage() >= budget.GetEstimatedBudget()) &&
+                    (blocksCount > static_cast<size_t>(GetMinimumBlockCount())) &&
+                    ((GetSize() - block->GetSize()) >= GetMinimumSize()))
+                {
+                    RemoveBlockAt(blockIndex);
+                }
+                else
+                {
+                    _emptyBlock = block;
+                }
+            }
+        }
+
+        IncrementallySortBlocks();
+    }
+
+    bool GraphicsMemoryBlockCollection::TryAllocate(size_t size,
+                                                    size_t alignment,
+                                                    GraphicsMemoryRegionAllocationFlags flags,
+                                                    GraphicsMemoryRegion<GraphicsMemoryBlock>& outRegion)
+    {
+        std::lock_guard<std::mutex> guard(_mutex); // TODO: come back to this when Tanner gets back to you
```

Comment:
Samesies - did Tanner get back to you?

Reply:
Yeah I had a discussion with him, but I think this ultimately boiled down to "leave it alone for now".